google-services-plugin/src/main/kotlin/com/google/gms/googleservices/GoogleServicesTask.kt

```diff
@@ -0,0 +1,345 @@
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.TaskAction
+
+@CacheableTask
+abstract class GoogleServicesTask : DefaultTask() {
+  @get:OutputDirectory abstract val outputDirectory: DirectoryProperty
+
+  @get:Internal
+  val intermediateDir: File
+    /** Reintroduced for binary compatiblity with the crashlytics plugin */
+    get() = outputDirectory.asFile.get()
+
+  @get:PathSensitive(PathSensitivity.RELATIVE)
+  @get:InputFiles
+  abstract val googleServicesJsonFiles: Property<Collection<File>>
+
+  @get:Input abstract val applicationId: Property<String>
+
+  @get:Input abstract val missingGoogleServicesStrategy: Property<MissingGoogleServicesStrategy>
+
+  @Throws(GradleException::class)
+  @TaskAction
+  fun action() {
+    val jsonFiles = googleServicesJsonFiles.get().filter { it.isFile }
+
+    if (jsonFiles.isEmpty()) {
+      val message =
+          """
+                File $JSON_FILE_NAME is missing. 
+                The Google Services Plugin cannot function without it. 
+                Searched locations: ${
+                googleServicesJsonFiles.get().joinToString { it.absolutePath }
+            }
+                """
+              .trimIndent()
+
+      when (missingGoogleServicesStrategy.get()) {
+        MissingGoogleServicesStrategy.ERROR -> throw GradleException(message)
+        MissingGoogleServicesStrategy.WARN -> logger.warn(message)
+        MissingGoogleServicesStrategy.IGNORE -> {
+          // ignore
+        }
+      }
+      return
+    }
+
+    val quickstartFile = jsonFiles.first()
+    logger.info("Parsing json file: " + quickstartFile.path)
+
+    // delete content of outputdir.
+    val intermediateDir = outputDirectory.get().asFile
+    intermediateDir.deleteRecursively()
+
+    if (!intermediateDir.mkdirs()) {
+      throw GradleException("Failed to create folder: $intermediateDir")
+    }
+
+    val root = JsonParser().parse(quickstartFile.bufferedReader(Charsets.UTF_8))
+    if (!root.isJsonObject) {
+      throw GradleException("Malformed root json")
+    }
+    val rootObject = root.asJsonObject
+    val resValues: MutableMap<String, String?> =
+        TreeMap() // TreeMap to preserve order with previous plugin versions
+    handleProjectNumberAndProjectId(rootObject, resValues)
+    handleFirebaseUrl(rootObject, resValues)
+    val clientObject = getClientForPackageName(rootObject)
+    if (clientObject != null) {
+      handleAnalytics(clientObject, resValues)
+      handleMapsService(clientObject, resValues)
+      handleGoogleApiKey(clientObject, resValues)
+      handleGoogleAppId(clientObject, resValues)
+      handleWebClientId(clientObject, resValues)
+    } else {
+      throw GradleException(
+          "No matching client found for package name '${applicationId.get()}' in ${quickstartFile.path}")
+    }
```

Comment:
No need to change here, but a more kotlin idiomatic way IMO woud be if we do something like

```
typealias FirebaseClientData = JsonObject
```

and `getClientForPackageName` would return that alias. Then all the inner "handle" methods would be implemented as extension functions of the alias, and we would get rid of `clientObject` params there.